Python/optimizer_analysis.c

```diff
@@ -9,10 +20,532 @@
+    for (int i = 0; i < curr_stackentries; i++) {
+        _Py_UOpsSymType *stackvar = sym_init_unknown(ctx);
+        if (stackvar == NULL) {
+            return NULL;
+        }
+        frame->stack[i] = stackvar;
+    }
+
+    return frame;
+}
+
+static void
+abstractcontext_fini(_Py_UOpsAbstractInterpContext *ctx)
+{
+    if (ctx == NULL) {
+        return;
+    }
+    ctx->curr_frame_depth = 0;
+    int tys = ctx->t_arena.ty_curr_number;
+    for (int i = 0; i < tys; i++) {
+        Py_CLEAR(ctx->t_arena.arena[i].const_val);
+    }
+}
+
+static int
+abstractcontext_init(
+    _Py_UOpsAbstractInterpContext *ctx,
+    PyCodeObject *co,
+    int curr_stacklen,
+    int ir_entries,
+    _PyUOpInstruction *new_writebuffer
+)
+{
+    ctx->limit = ctx->locals_and_stack + MAX_ABSTRACT_INTERP_SIZE;
+    ctx->n_consumed = ctx->locals_and_stack;
+#ifdef Py_DEBUG // Aids debugging a little. There should never be NULL in the abstract interpreter.
+    for (int i = 0 ; i < MAX_ABSTRACT_INTERP_SIZE; i++) {
+        ctx->locals_and_stack[i] = NULL;
+    }
+#endif
+
+    // Setup the arena for sym expressions.
+    ctx->t_arena.ty_curr_number = 0;
+    ctx->t_arena.ty_max_number = TY_ARENA_SIZE;
+
+    // Frame setup
+    ctx->curr_frame_depth = 0;
+    _Py_UOpsAbstractFrame *frame = ctx_frame_new(ctx, co, ctx->n_consumed, 0, curr_stacklen);
+    if (frame == NULL) {
+        return -1;
+    }
+    ctx->curr_frame_depth++;
+    ctx->frame = frame;
+
+    // IR and sym setup
+    ctx->frequent_syms.push_nulL_sym = NULL;
+
+    // Emitter setup
+    ctx->emitter.writebuffer = new_writebuffer;
+    ctx->emitter.curr_i = 0;
+    ctx->emitter.writebuffer_end = new_writebuffer + ir_entries;
+    return 0;
+}
+
+
+static int
+ctx_frame_pop(
+    _Py_UOpsAbstractInterpContext *ctx
+)
+{
+    _Py_UOpsAbstractFrame *frame = ctx->frame;
+
+    ctx->n_consumed = frame->locals;
+    ctx->curr_frame_depth--;
+    assert(ctx->curr_frame_depth >= 1);
+    ctx->frame = &ctx->frames[ctx->curr_frame_depth - 1];
+
+    return 0;
+}
+
+static void
+sym_set_type_from_const(_Py_UOpsSymType *sym, PyObject *obj);
+
+// Steals a reference to const_val
+static _Py_UOpsSymType*
+_Py_UOpsSymType_New(_Py_UOpsAbstractInterpContext *ctx,
+                               PyObject *const_val)
+{
+    _Py_UOpsSymType *self = &ctx->t_arena.arena[ctx->t_arena.ty_curr_number];
+    if (ctx->t_arena.ty_curr_number >= ctx->t_arena.ty_max_number) {
+        OPT_STAT_INC(optimizer_failure_reason_no_memory);
+        DPRINTF(1, "out of space for symbolic expression type\n");
+        return NULL;
+    }
+    ctx->t_arena.ty_curr_number++;
+    self->const_val = NULL;
+    self->types = 0;
+
+    if (const_val != NULL) {
+        self->const_val = Py_NewRef(const_val);
+    }
+
+    return self;
+}
+
+
+static void
+sym_set_type(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ)
+{
+    sym->types |= 1 << typ;
+}
+
+
+static _Py_UOpsSymExprTypeEnum
+pytype_to_type(PyTypeObject *tp)
```

Comment:
If I call this with `&Bool_Type` it will fail.
This will make handling of `TO_BOOL` specializations unnecessarily difficult.
It would be more robust if the actual `PyTypeObject *` pointer were stored.
We will also want to track other types like builtin functions, lists, sets, tuples, etc.